### ISSUE TITLE
🐛 Bug Report — `workerd` sometimes doesn't send `Debugger.scriptParsed` events

### ISSUE THREAD

**Issue by mrbbot** [cloudflare/workerd#1201]
Hey! 👋 When enabling the inspector with the hello world ESM sample, `workerd` can start up in an invalid state where it never responds to `Debugger.enable` commands with `Debugger.scriptParsed` events.

Using the following script and `workerd@1.20230914.0` or `workerd@1.20230918.0`...

```mjs
// client.mjs
import { WebSocket } from "ws";

const ws = new WebSocket("ws://127.0.0.1:9229/main");

ws.addEventListener("message", (event) => {
   console.log(JSON.parse(event.data));
});
ws.addEventListener("open", (event) => {
    ws.send(JSON.stringify({ id: 0, method: "Debugger.enable", params: {} }));
    setTimeout(() => {
        ws.close();
    }, 3000);
});
```

```
$ npx workerd@1.20230914.0 serve samples/helloworld_esm/config.capnp --verbose --inspector-addr=127.0.0.1:9229
workerd/server/server.c++:2622: info: Inspector is listening
workerd/server/server.c++:1099: info: Inspector client attaching [main]
workerd/server/server.c++:1109: info: Inspector client detached [(��]
```

```
$ npm install ws && node client.mjs
// Expect `Debugger.scriptParsed` events here
{
  id: 0,
  result: { debuggerId: '-7930107370981718434.8415374541215293299' }
}
```

You may need to re-run `npx workerd@... serve` to get it into an invalid state. I haven't been able to reproduce this with `workerd@1.20230904.0`. Also note the garbage in the inspector detach logs `[(��]`. 

I haven't been able to narrow this down to a particular change yet as my local `workerd` builds are currently broken because of an automatic Xcode update. I'm also going OOO for the next 7 days.

**Comment by ohodson** [comment]
I can repro this with-or-without the change in the initial execution context id that workerd added to v8.

The repro is fairly easy: launch `workerd` with the helloworld sample (the default debug target in VSCode for workerd with inspector enabled). Simply connecting and disconnecting inspector clients from Chrome shows roughly 1 in 3 not getting the `Runtime.executionContextCreated` and `Debugger.scriptParsed` messages.

I also came across a Smi initialization exception with the initial execution context id randomization. This is an easy to fix we should make, but not relevant to the problem at hand.

**Comment by ohodson** [comment]
Bisecting:

| Commit | Description | Repro?  |
| - | - |
| 3f8ecc87 | V8 update to 11.7 | Yes, plus Smi exception |
| 5166892d | Commit before (772bf8d1 randomize id change) | |

**Comment by ohodson** [comment]
Take 2, just linux and Chrome

Commit (most recent first) | Description | Repro?
| --: | -- | -- |
| 6b19807d6 | Update V8 to 11.6 | |

**Comment by ohodson** [comment]
Adding a pause before sending messages appears to enable Chrome to render the CDP messages it receives in the protocol view window. tcpdump shows those same messages arriving and being responded to if there is no delay, but the UX does not render it.

There might a better fix available, TBD.

**Comment by ohodson** [comment]
Overall, it'd be good to have another fix here, but this issue seems to be pressing this week.

On both the local M1 MBP and a x64 Linux box, 250ms seems to eliminate any issues when testing `workerd/samples/helloworld/config.capnp` against Chrome's protocol inspector (and what's observed in the developer UX). 300ms is in the PR as it adds some margin for safety. 50ms, 100ms, 150ms, 200ms all exhibited some degree of flakiness on both test boxes.

**Comment by ohodson** [comment]
Looking at the CDP, it looks like the initial messages are generated appropriately.

1. Client sends [Runtime.enable](https://chromedevtools.github.io/devtools-protocol/tot/Runtime/#method-enable), this enables the reporting of execution context creation by [Runtime.executionContextCreated](https://chromedevtools.github.io/devtools-protocol/tot/Runtime/#event-executionContextCreated) events.
2. Client sends [Debugger.enable](https://chromedevtools.github.io/devtools-protocol/tot/Debugger/#method-enable) and this generates [Debugger.scriptParsed](https://chromedevtools.github.io/devtools-protocol/tot/Debugger/#event-scriptParsed) events. The wording for this in the description of `Debugger.scriptParsed`: _... This event is also fired for all known and uncollected scripts upon enabling debugger_.

In wrangler-devtools, It looks like `scriptParsed` events are ignored if the debugger is not enabled (at least in `DebuggerModel.ts`). However, I'm not seeing any similar gating for `executionContextCreated` events in `RuntimeModel.ts` (and I'm not at all familiar with this code base).

**Comment by ohodson** [comment]
Reset on this issue.

With an `opt` workerd build at tag `v1.20230918.0`, the problem as described in the initial report here repros on my MBP.

Repro steps:
* workerd launched from vscode in `workerd with inspector enabled (opt)` with the default sample `${workspaceFolder}/samples/helloworld/config.capnp`
* `node client.mjs`

Observations:
* Zero successes from invoking the client script.
* Adding a `KJ_LOG` on the receive path and the issue goes away.
* No repro with a `dbg` build.

Follow-on:
* The issue didn't repro at ToT with-or-without yesterday's change (ad540fb826 nor 07c50e35b74f17).
* Going back to tag `v1.20230918.0`, rebuilding `opt` workerd, and no repro.

Interesting.

**Comment by ohodson** [comment]
💡Forcing the smi issue fixed in #1217 with an opt build and this issue appears to repro 100% at `v1.20230918.0`.
* There is no output from `workerd`.
* The executionContextId is bumped by 1 each time a client connects, but if the initial choice was bad, no client is likely to a response.
* The smi issue shows up in a debug only check (`V8_Dcheck`).

**Comment by RamIdeas** [comment]
@ohodson am I inferring correctly that:

[1] this bug was caused by the RNG of executionContextId (but we're not just not clear why), and
[2] the workaround to wait 300ms before sending CDP messages is sufficient for now so users won't be affected by the bug

I'm sure what I've inferred is probably wrong but I'm just seeking clarity on whether we're happy to release workerd in it's current state or what the latest build we're comfortable releasing is

**Comment by ohodson** [comment]
Just to follow up:

1) the RNG bug for `executionContextId` happened because the 32-bit integer field that holds the last executionContextId (and the initial RNG value) gets converted into a small integer in V8 before being put into an inspector protocol message. A SMI is not large enough to span the full 32-bit range (varies by CPU architecture, IIRC, but is typically 30-31 bits). This showed up as a problem when testing with a dbg build. The failure is non-deterministic and we don't have a test that would catch it. It is fixed in #1217.

2) the 300ms appears to be necessary for chrome devtools to render the initial CDP messages between devtools and workerd/v8. It appears they don't get rendered if they occur before the creation of the devtools window. I think we can revert this, but being able to see those messages also saves time when looking into problems like this (otherwise have to pull out wireshark to see them or add extra instrumentation to workerd). If it's too annoying for wrangler, let's just revert.
